src/api/uploads/BaseUpload.js

```diff
@@ -60,11 +69,51 @@ class BaseUpload extends Base {
         this.xhr.options({
             url,
             data: attributes,
-            successHandler: this.preflightSuccessHandler,
+            successHandler: this.preflightSuccessReachabilityHandler,
             errorHandler: this.preflightErrorHandler,
         });
     };
 
+    /**
+     * Handles successful preflight response.
+     * Performs a upload reachability test before calling preflightSuccessHandler.
+     *
+     * @param {Object} - Request options
+     * @return {Promise} Async function promise
+     */
+    preflightSuccessReachabilityHandler = async ({ data }: { data: { upload_url?: string } }): Promise<any> => {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        const uploadUrl = data.upload_url;
+        // If uploadUrl is not avaliable, don't make reachability test
+        if (!uploadUrl) {
+            this.preflightSuccessHandler({ data });
+            return;
+        }
+
+        // Parse upload host from uploadUrl
+        const splitUrl = uploadUrl.split('/');
+        const uploadHost = `${splitUrl[0]}//${splitUrl[2]}/`;
+        // The default upload host should always reachable
+        if (uploadHost === `${DEFAULT_HOSTNAME_UPLOAD}/`) {
+            this.preflightSuccessHandler({ data });
+            return;
+        }
+
+        // If upload host reachable upload file, else make a new preflight request
+        const isHostReachable = await this.uploadsReachability.isReachable(uploadHost);
+        if (isHostReachable) {
+            this.preflightSuccessHandler({ data });
+        } else if (this.reachabilityRetryCount >= MAX_REACHABILITY_RETRY) {
+            this.preflightSuccessHandler();
+        } else {
+            this.reachabilityRetryCount += 1;
+            this.makePreflightRequest();
```

Comment:
so i'm assuming we are retrying with another preflight request because we got back an unreachable host is that right? should we be expecting that the preflight request would give us back a different upload url with each call?

i would actually say that if we got an unreachable host back from the preflight response, we might consider that a preflight error and just call preflightErrorHandler which handles the retry logic and calls makePreflightRequest. that would eliminate the need to track reachabilityRetries.

what are your thoughts?

Reply:
> so i'm assuming we are retrying with another preflight request because we got back an unreachable host is that right?

yes

> should we be expecting that the preflight request would give us back a different upload url with each call?

Yes, the preflight request would give us a different upload url with each call given that we keep updating the `unreachable_hosts` query parameter.

I do like the idea of calling the preflightErrorHandler if the preflight returns an unreachable host. But we will have to increase the `MAX_RETRY` value to a higher number. This might cause some problems since we will retry more for other scenarios. For example, if we are being rate limited 429, we might end up retrying more than originally intended. 
I also don't like the idea of completely canceling the upload if we reach max retries due to reachability tests failing. We are always adding more upload hosts, so the list of possible hosts will keep growing over time. This will cause us to increase the overall `MAX_RETRY` count which like I mentioned above might lead to some unintended consequences. If we reach the max retry count due to reachability tests failing we should fallback to the default upload hosts, which should always be reachable, instead of canceling the upload. This is why I decided to keep the `reachabilityRetryCount` separate from `retryCount`. 

Let me know what you think.